File: python/2349. Design a Number Container System.py
class NumberContainers:

    def __init__(self):
        self.index2number = {}
        self.number2heap = {}

    def change(self, index: int, number: int) -> None:
        self.index2number[index] = number
        if number in self.number2heap:
            heapq.heappush(self.number2heap[number], index)
        else:
            self.number2heap[number] = [index]
        

    def find(self, number: int) -> int:
        if number not in self.number2heap:
            return -1
        else:
            heap = self.number2heap[number]
            while heap:
                index = heap[0]
                if self.index2number.get(index, -1) == number:
                    return index
                else:
                    heapq.heappop(heap)
            return -1
        


# Your NumberContainers object will be instantiated and called as such:
# obj = NumberContainers()
# obj.change(index,number)
# param_2 = obj.find(number)

File: rust/2349. Design a Number Container System.rs
struct NumberContainers {
    index2number: std::collections::HashMap<i32, i32>,
    number2heap: std::collections::HashMap<i32, std::collections::BinaryHeap<i32>>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl NumberContainers {
    fn new() -> Self {
        NumberContainers {
            index2number: std::collections::HashMap::new(),
            number2heap: std::collections::HashMap::new(),
        }
    }
    
    fn change(&mut self, index: i32, number: i32) {
        self.index2number.insert(index, number);
        self.number2heap
            .entry(number)
            .or_insert_with(std::collections::BinaryHeap::new)
            .push(-index);
    }
    
    fn find(&mut self, number: i32) -> i32 {
        self.number2heap.get_mut(&number).and_then(|heap| {
            while let Some(&index) = heap.peek() {
                if self.index2number.get(&-index) == Some(&number) {
                    return Some(-index);
                }
                heap.pop();
            }
            None
        }).unwrap_or(-1)
    }
}

/**
 * Your NumberContainers object will be instantiated and called as such:
 * let obj = NumberContainers::new();
 * obj.change(index, number);
 * let ret_2: i32 = obj.find(number);
 */
